File: tests/cpp/test_conservative_interpolate.cpp
//  ██████   ██████ ██████████  █████████  █████   █████ █████    ███████
// ░░██████ ██████ ░░███░░░░░█ ███░░░░░███░░███   ░░███ ░░███   ███░░░░░███      ███         ███
//  ░███░█████░███  ░███  █ ░ ░███    ░░░  ░███    ░███  ░███  ███     ░░███    ░███        ░███
//  ░███░░███ ░███  ░██████   ░░█████████  ░███████████  ░███ ░███      ░███ ███████████ ███████████
//  ░███ ░░░  ░███  ░███░░█    ░░░░░░░░███ ░███░░░░░███  ░███ ░███      ░███░░░░░███░░░ ░░░░░███░░░
//  ░███      ░███  ░███ ░   █ ███    ░███ ░███    ░███  ░███ ░░███     ███     ░███        ░███
//  █████     █████ ██████████░░█████████  █████   █████ █████ ░░░███████░      ░░░         ░░░
// ░░░░░     ░░░░░ ░░░░░░░░░░  ░░░░░░░░░  ░░░░░   ░░░░░ ░░░░░    ░░░░░░░
//
//
//  License:         MIT License
//                   meshio++ default license: LICENSE
//
//  Main authors:    Vicente Mataix Ferrandiz
//
//

// System includes
#include <algorithm>
#include <cmath>
#include <cstddef>
#include <cstdint>
#include <stdexcept>
#include <string>
#include <vector>

// External includes
#include <gtest/gtest.h>

// Project includes
#include "mesh_fixtures.hpp"
#include "meshioplusplus/detail/value_io.hpp"
#include "meshioplusplus/operations/conservative_interpolate.hpp"

namespace {

using meshioplusplus::conservative_interpolate;
using meshioplusplus::conservative_interpolate_conflict_from_name;
using meshioplusplus::ConservativeInterpolateConflict;
using meshioplusplus::ConservativeInterpolateOptions;
using meshioplusplus::Mesh;
using meshioplusplus::NDArray;
namespace md = meshioplusplus::detail;

// --- fixtures ----------------------------------------------------------------

// A single triangle in the z = 0 plane with corners `p0`, `p1`, `p2`.
Mesh one_triangle(const std::vector<double>& p0, const std::vector<double>& p1,
                  const std::vector<double>& p2) {
    return mt::make_mesh({p0, p1, p2}, "triangle", {{0, 1, 2}});
}

// A single tetrahedron with the given 4 corners.
Mesh one_tetra(const std::vector<double>& p0, const std::vector<double>& p1,
               const std::vector<double>& p2, const std::vector<double>& p3) {
    return mt::make_mesh({p0, p1, p2, p3}, "tetra", {{0, 1, 2, 3}});
}

// An n x n grid of unit quads in the z = 0 plane, scaled to cover
// [0, n * scale] x [0, n * scale].
Mesh quad_grid(int n, double scale) {
    const int m = n + 1;
    std::vector<std::vector<double>> pts;
    for (int j = 0; j < m; ++j)
        for (int i = 0; i < m; ++i)
            pts.push_back({static_cast<double>(i) * scale, static_cast<double>(j) * scale, 0.0});
    auto pid = [m](int i, int j) { return static_cast<std::int64_t>(j * m + i); };
    std::vector<std::vector<std::int64_t>> cells;
    for (int j = 0; j < n; ++j)
        for (int i = 0; i < n; ++i)
            cells.push_back({pid(i, j), pid(i + 1, j), pid(i + 1, j + 1), pid(i, j + 1)});
    return mt::make_mesh(std::move(pts), "quad", std::move(cells));
}

// An n x n x n grid of unit hexahedra, scaled to cover [0, n*scale]^3.
Mesh hex_grid(int n, double scale) {
    const int m = n + 1;
    std::vector<std::vector<double>> pts;
    for (int k = 0; k < m; ++k)
        for (int j = 0; j < m; ++j)
            for (int i = 0; i < m; ++i)
                pts.push_back({static_cast<double>(i) * scale, static_cast<double>(j) * scale,
                               static_cast<double>(k) * scale});
    auto pid = [m](int i, int j, int k) { return static_cast<std::int64_t>((k * m + j) * m + i); };
    std::vector<std::vector<std::int64_t>> cells;
    for (int k = 0; k < n; ++k)
        for (int j = 0; j < n; ++j)
            for (int i = 0; i < n; ++i)
                cells.push_back({pid(i, j, k), pid(i + 1, j, k), pid(i + 1, j + 1, k),
                                 pid(i, j + 1, k), pid(i, j, k + 1), pid(i + 1, j, k + 1),
                                 pid(i + 1, j + 1, k + 1), pid(i, j + 1, k + 1)});
    return mt::make_mesh(std::move(pts), "hexahedron", std::move(cells));
}

// Total sum(value * measure) over a mesh's cell_data array `name`, weighting
// each quad/hexahedron cell by its own area/volume computed independently of
// the operation under test (a plain shoelace/box-volume formula).
double weighted_sum(const Mesh& rMesh, const std::string& name) {
    double total = 0.0;
    std::size_t block = 0;
    for (const auto cb : rMesh.CellRange()) {
        const NDArray& data = rMesh.CellData(name, block);
        const NDArray& conn = cb.Conn();
        const NDArray& pts = rMesh.Points();
        const std::size_t pdim = rMesh.PointDim();
        const std::size_t npc = cb.NodesPerCell();
        for (std::size_t c = 0; c < cb.NumCells(); ++c) {
            std::vector<double> xs, ys, zs;
            for (std::size_t k = 0; k < npc; ++k) {
                const std::int64_t node = md::read_int(conn, c * npc + k);
                xs.push_back(md::read_double(pts, static_cast<std::size_t>(node) * pdim + 0));
                ys.push_back(md::read_double(pts, static_cast<std::size_t>(node) * pdim + 1));
                if (pdim > 2)
                    zs.push_back(md::read_double(pts, static_cast<std::size_t>(node) * pdim + 2));
            }
            double measure;
            if (npc == 4) {
                // shoelace area of the (planar, z ignored) quad
                double a2 = 0.0;
                for (std::size_t k = 0; k < 4; ++k) {
                    const std::size_t k2 = (k + 1) % 4;
                    a2 += xs[k] * ys[k2] - xs[k2] * ys[k];
                }
                measure = std::abs(a2) * 0.5;
            } else {
                // axis-aligned hexahedron volume from its bbox
                const double dx = *std::max_element(xs.begin(), xs.end()) -
                                  *std::min_element(xs.begin(), xs.end());
                const double dy = *std::max_element(ys.begin(), ys.end()) -
                                  *std::min_element(ys.begin(), ys.end());
                const double dz = *std::max_element(zs.begin(), zs.end()) -
                                  *std::min_element(zs.begin(), zs.end());
                measure = dx * dy * dz;
            }
            total += measure * md::read_double(data, c);
        }
        ++block;
    }
    return total;
}

// --- clip correctness, via single-cell meshes (isolates the kernel) --------

TEST(ConservativeInterpolate, IdenticalTriangleRecoversFullValue) {
    Mesh src = one_triangle({0, 0, 0}, {1, 0, 0}, {0, 1, 0});
    src.AddCellData("f", {mt::data_array({7.0})});
    Mesh tgt = one_triangle({0, 0, 0}, {1, 0, 0}, {0, 1, 0});

    const Mesh out = conservative_interpolate(src, tgt);
    EXPECT_DOUBLE_EQ(md::read_double(out.CellData("f", 0), 0), 7.0);
}

TEST(ConservativeInterpolate, IdenticalTetraRecoversFullValue) {
    Mesh src = one_tetra({0, 0, 0}, {1, 0, 0}, {0, 1, 0}, {0, 0, 1});
    src.AddCellData("f", {mt::data_array({3.5})});
    Mesh tgt = one_tetra({0, 0, 0}, {1, 0, 0}, {0, 1, 0}, {0, 0, 1});

    const Mesh out = conservative_interpolate(src, tgt);
    EXPECT_DOUBLE_EQ(md::read_double(out.CellData("f", 0), 0), 3.5);
}

TEST(ConservativeInterpolate, DisjointTrianglesGiveDefaultValue) {
    Mesh src = one_triangle({0, 0, 0}, {1, 0, 0}, {0, 1, 0});
    src.AddCellData("f", {mt::data_array({9.0})});
    Mesh tgt = one_triangle({10, 10, 0}, {11, 10, 0}, {10, 11, 0});

    ConservativeInterpolateOptions opts;
    opts.mDefaultValue = -1.0;
    const Mesh out = conservative_interpolate(src, tgt, opts);
    EXPECT_DOUBLE_EQ(md::read_double(out.CellData("f", 0), 0), -1.0);
}

TEST(ConservativeInterpolate, DisjointTetrasGiveDefaultValue) {
    Mesh src = one_tetra({0, 0, 0}, {1, 0, 0}, {0, 1, 0}, {0, 0, 1});
    src.AddCellData("f", {mt::data_array({9.0})});
    Mesh tgt = one_tetra({10, 10, 10}, {11, 10, 10}, {10, 11, 10}, {10, 10, 11});

    ConservativeInterpolateOptions opts;
    opts.mDefaultValue = -1.0;
    const Mesh out = conservative_interpolate(src, tgt, opts);
    EXPECT_DOUBLE_EQ(md::read_double(out.CellData("f", 0), 0), -1.0);
}

// A target triangle exactly half of the source triangle's area (the source
// right triangle's mid-line split) -> the overlap covers the target
// entirely, so the target's own value is exactly the source's.
TEST(ConservativeInterpolate, PartialTriangleOverlapRecoversSourceValueWhenFullyCovered) {
    Mesh src = one_triangle({0, 0, 0}, {2, 0, 0}, {0, 2, 0});
    src.AddCellData("f", {mt::data_array({4.0})});
    Mesh tgt = one_triangle({0, 0, 0}, {1, 0, 0}, {0, 1, 0});  // strictly inside src

    const Mesh out = conservative_interpolate(src, tgt);
    EXPECT_DOUBLE_EQ(md::read_double(out.CellData("f", 0), 0), 4.0);
}

TEST(ConservativeInterpolate, PartialTetraOverlapRecoversSourceValueWhenFullyCovered) {
    Mesh src = one_tetra({0, 0, 0}, {2, 0, 0}, {0, 2, 0}, {0, 0, 2});
    src.AddCellData("f", {mt::data_array({6.0})});
    Mesh tgt = one_tetra({0, 0, 0}, {1, 0, 0}, {0, 1, 0}, {0, 0, 1});  // strictly inside src

    const Mesh out = conservative_interpolate(src, tgt);
    EXPECT_DOUBLE_EQ(md::read_double(out.CellData("f", 0), 0), 6.0);
}

// Two triangles overlapping in a known quadrilateral: src is the unit square
// (as two triangles is not needed -- a single big triangle) clipped against a
// target triangle offset by (0.5, 0.5) with corners chosen so the exact
// overlap area is computable by hand: src = right triangle (0,0)-(2,0)-(0,2),
// tgt = right triangle (0,0)-(2,0)-(0,2) translated and reflected so the
// overlap is the square [0,1]x[0,1] intersected with src (a known triangle
// of area 0.5). Use a constant-source-value trick: a uniform field of value
// V over src, sampled with a target that partially overlaps by a KNOWN
// fraction, must read back exactly V (weighted mean of a constant is that
// constant) -- so exact-area verification is instead done through the
// integral-conservation tests below, which do not depend on knowing the
// overlap area of any single pair.

// --- operation-level behaviour ----------------------------------------------

TEST(ConservativeInterpolate, CellDataConservesTotalIntegral2D) {
    // Two independent partitions of the exact same [0,6]x[0,6] domain.
    Mesh src = quad_grid(3, 2.0);  // 3x3 cells of size 2
    Mesh tgt = quad_grid(4, 1.5);  // 4x4 cells of size 1.5

    std::vector<double> vals;
    for (int j = 0; j < 3; ++j)
        for (int i = 0; i < 3; ++i)
            vals.push_back(1.0 + 2.0 * i + 3.0 * j);  // a varying, non-constant field
    src.AddCellData("f", {mt::data_array(vals)});

    const Mesh out = conservative_interpolate(src, tgt);
    EXPECT_NEAR(weighted_sum(out, "f"), weighted_sum(src, "f"), 1e-8);
}

TEST(ConservativeInterpolate, CellDataConservesTotalIntegral3D) {
    Mesh src_hex = hex_grid(2, 3.0);  // 2x2x2 cells of size 3, domain [0,6]^3
    Mesh tgt_hex = hex_grid(3, 2.0);  // 3x3x3 cells of size 2, domain [0,6]^3

    std::vector<double> vals;
    for (int i = 0; i < 8; ++i)
        vals.push_back(1.0 + static_cast<double>(i));
    src_hex.AddCellData("f", {mt::data_array(vals)});

    const Mesh out = conservative_interpolate(src_hex, tgt_hex);
    EXPECT_NEAR(weighted_sum(out, "f"), weighted_sum(src_hex, "f"), 1e-6);
}

TEST(ConservativeInterpolate, RejectsMismatchedTopologicalDimension) {
    Mesh src2d = one_triangle({0, 0, 0}, {1, 0, 0}, {0, 1, 0});
    src2d.AddCellData("f", {mt::data_array({1.0})});
    Mesh tgt3d = one_tetra({0, 0, 0}, {1, 0, 0}, {0, 1, 0}, {0, 0, 1});
    EXPECT_THROW(conservative_interpolate(src2d, tgt3d), std::invalid_argument);
}

TEST(ConservativeInterpolate, PresentInBothLocationsTransfersBoth) {
    Mesh src = one_triangle({0, 0, 0}, {1, 0, 0}, {0, 1, 0});
    src.AddCellData("f", {mt::data_array({2.0})});
    src.AddPointData("f", mt::data_array({1.0, 2.0, 3.0}));
    Mesh tgt = one_triangle({0, 0, 0}, {1, 0, 0}, {0, 1, 0});

    const Mesh out = conservative_interpolate(src, tgt);
    EXPECT_TRUE(out.HasCellData("f"));
    EXPECT_TRUE(out.HasPointData("f"));
}

TEST(ConservativeInterpolate, ConflictErrorOverwriteSuffix) {
    Mesh src = one_triangle({0, 0, 0}, {1, 0, 0}, {0, 1, 0});
    src.AddCellData("f", {mt::data_array({2.0})});
    Mesh tgt = one_triangle({0, 0, 0}, {1, 0, 0}, {0, 1, 0});
    tgt.AddCellData("f", {mt::data_array({99.0})});

    EXPECT_THROW(conservative_interpolate(src, tgt), std::invalid_argument);

    ConservativeInterpolateOptions overwrite_opts;
    overwrite_opts.mOnConflict = ConservativeInterpolateConflict::Overwrite;
    const Mesh out_ow = conservative_interpolate(src, tgt, overwrite_opts);
    EXPECT_DOUBLE_EQ(md::read_double(out_ow.CellData("f", 0), 0), 2.0);

    ConservativeInterpolateOptions suffix_opts;
    suffix_opts.mOnConflict = ConservativeInterpolateConflict::Suffix;
    const Mesh out_sfx = conservative_interpolate(src, tgt, suffix_opts);
    EXPECT_TRUE(out_sfx.HasCellData("f_interp"));
    EXPECT_DOUBLE_EQ(md::read_double(out_sfx.CellData("f", 0), 0), 99.0);
}

TEST(ConservativeInterpolate, TargetGeometryAndOwnDataUntouched) {
    Mesh src = one_triangle({0, 0, 0}, {1, 0, 0}, {0, 1, 0});
    src.AddCellData("f", {mt::data_array({2.0})});
    Mesh tgt = one_triangle({0, 0, 0}, {1, 0, 0}, {0, 1, 0});
    tgt.AddCellData("g", {mt::data_array({42.0})});

    const Mesh out = conservative_interpolate(src, tgt);
    EXPECT_EQ(out.NumPoints(), tgt.NumPoints());
    EXPECT_DOUBLE_EQ(md::read_double(out.CellData("g", 0), 0), 42.0);
}

TEST(ConservativeInterpolate, UnknownArrayThrows) {
    Mesh src = one_triangle({0, 0, 0}, {1, 0, 0}, {0, 1, 0});
    Mesh tgt = one_triangle({0, 0, 0}, {1, 0, 0}, {0, 1, 0});
    ConservativeInterpolateOptions opts;
    opts.mArrays = {"nope"};
    EXPECT_THROW(conservative_interpolate(src, tgt, opts), std::invalid_argument);
}

TEST(ConservativeInterpolate, EmptySourceThrows) {
    Mesh src;
    Mesh tgt = one_triangle({0, 0, 0}, {1, 0, 0}, {0, 1, 0});
    EXPECT_THROW(conservative_interpolate(src, tgt), std::invalid_argument);
}

TEST(ConservativeInterpolate, PointDataComposedConservesApproximately) {
    Mesh src = quad_grid(3, 2.0);
    Mesh tgt = quad_grid(4, 1.5);
    std::vector<double> vals;
    for (int j = 0; j < 4; ++j)
        for (int i = 0; i < 4; ++i)
            vals.push_back(1.0 + i + j);
    src.AddPointData("f", mt::data_array(vals));

    const Mesh out = conservative_interpolate(src, tgt);
    ASSERT_TRUE(out.HasPointData("f"));
    // Approximate: point<->cell lumping on both sides is lossy, but the
    // result must be a real, finite, in-range value -- not NaN/garbage.
    const NDArray& f = out.PointData("f");
    for (std::size_t i = 0; i < out.NumPoints(); ++i) {
        const double v = md::read_double(f, i);
        EXPECT_TRUE(std::isfinite(v));
    }
}

TEST(ConservativeInterpolate, DuplicateCandidatesFromMultiCellBboxAreNotDoubleCounted) {
    // Two adjacent, DIFFERENTLY-valued source quads and one target quad that
    // exactly spans both. A duplicated or dropped candidate from the
    // broad-phase box query would skew the ratio away from the exact
    // equal-area-weighted average -- unlike a single-source-cell scenario,
    // where over-counting cancels out of the numerator/denominator ratio and
    // so cannot be detected this way.
    Mesh src = mt::make_mesh({{0, 0, 0}, {1, 0, 0}, {1, 1, 0}, {0, 1, 0}, {2, 0, 0}, {2, 1, 0}},
                             "quad", {{0, 1, 2, 3}, {1, 4, 5, 2}});
    src.AddCellData("f", {mt::data_array({10.0, 30.0})});
    Mesh tgt = mt::make_mesh({{0, 0, 0}, {2, 0, 0}, {2, 1, 0}, {0, 1, 0}}, "quad", {{0, 1, 2, 3}});

    const Mesh out = conservative_interpolate(src, tgt);
    EXPECT_NEAR(md::read_double(out.CellData("f", 0), 0), 20.0, 1e-9);
}

}  // namespace

File: tests/python/test_conservative_interpolate.py
"""Tests for mass-preserving cross-mesh field transfer
(the ``conservative_interpolate`` operation)."""

import numpy as np
import pytest

import meshioplusplus
from meshioplusplus import conservative_interpolate

try:
    from meshioplusplus import _core
except ImportError:  # pragma: no cover - a pure-Python build
    _core = None

needs_core = pytest.mark.skipif(
    _core is None, reason="conservative_interpolate has no pure-Python fallback"
)


def _quad_grid(n, scale):
    """An n x n grid of quads over [0, n*scale]^2."""
    m = n + 1
    pts = np.array(
        [[i * scale, j * scale, 0.0] for j in range(m) for i in range(m)], float
    )

    def pid(i, j):
        return j * m + i

    cells = np.array(
        [
            [pid(i, j), pid(i + 1, j), pid(i + 1, j + 1), pid(i, j + 1)]
            for j in range(n)
            for i in range(n)
        ]
    )
    return meshioplusplus.Mesh(pts, [("quad", cells)])


def _hex_grid(n, scale):
    """An n x n x n grid of hexahedra over [0, n*scale]^3."""
    m = n + 1
    pts = []
    for k in range(m):
        for j in range(m):
            for i in range(m):
                pts.append([i * scale, j * scale, k * scale])
    pts = np.array(pts, float)

    def pid(i, j, k):
        return (k * m + j) * m + i

    cells = []
    for k in range(n):
        for j in range(n):
            for i in range(n):
                cells.append(
                    [
                        pid(i, j, k),
                        pid(i + 1, j, k),
                        pid(i + 1, j + 1, k),
                        pid(i, j + 1, k),
                        pid(i, j, k + 1),
                        pid(i + 1, j, k + 1),
                        pid(i + 1, j + 1, k + 1),
                        pid(i, j + 1, k + 1),
                    ]
                )
    return meshioplusplus.Mesh(pts, [("hexahedron", np.array(cells))])


def _quad_weighted_sum(mesh, name):
    total = 0.0
    for block, data in zip(mesh.cells, mesh.cell_data[name]):
        conn = np.asarray(block.data)
        pts = np.asarray(mesh.points)
        for c in range(len(conn)):
            xs = pts[conn[c], 0]
            ys = pts[conn[c], 1]
            a2 = 0.0
            for k in range(4):
                k2 = (k + 1) % 4
                a2 += xs[k] * ys[k2] - xs[k2] * ys[k]
            total += abs(a2) / 2.0 * np.asarray(data)[c]
    return total


def _one_triangle(p0, p1, p2):
    return meshioplusplus.Mesh(
        np.array([p0, p1, p2], float), [("triangle", np.array([[0, 1, 2]]))]
    )


def _one_polyhedron_tet(p0, p1, p2, p3):
    """A single tetrahedron stored as a polyhedron block, to exercise the
    "ragged/polyhedron accepted for free" scope decision."""
    pts = np.array([p0, p1, p2, p3], float)
    faces = [[0, 1, 3], [1, 2, 3], [2, 0, 3], [0, 2, 1]]
    return meshioplusplus.Mesh(pts, [("polyhedron4", [faces])])


@needs_core
def test_cell_data_conserves_total_integral_across_different_partitions():
    src = _quad_grid(3, 2.0)
    tgt = _quad_grid(4, 1.5)
    src.cell_data["f"] = [
        np.array([1.0 + 2 * i + 3 * j for j in range(3) for i in range(3)])
    ]

    out = conservative_interpolate(src, tgt)
    assert _quad_weighted_sum(out, "f") == pytest.approx(
        _quad_weighted_sum(src, "f"), abs=1e-8
    )


@needs_core
def test_point_data_is_transferred_via_composition():
    src = _quad_grid(3, 2.0)
    tgt = _quad_grid(4, 1.5)
    src.point_data["T"] = np.array([1.0 + i + j for j in range(4) for i in range(4)])

    out = conservative_interpolate(src, tgt)
    assert "T" in out.point_data
    assert np.all(np.isfinite(out.point_data["T"]))


@needs_core
def test_present_in_both_locations_transfers_both():
    src = _one_triangle([0, 0, 0], [1, 0, 0], [0, 1, 0])
    src.cell_data["f"] = [np.array([2.0])]
    src.point_data["f"] = np.array([1.0, 2.0, 3.0])
    tgt = _one_triangle([0, 0, 0], [1, 0, 0], [0, 1, 0])

    out = conservative_interpolate(src, tgt)
    assert "f" in out.cell_data
    assert "f" in out.point_data


@needs_core
def test_default_value_fills_uncovered_cells():
    src = _one_triangle([0, 0, 0], [1, 0, 0], [0, 1, 0])
    src.cell_data["f"] = [np.array([9.0])]
    tgt = _one_triangle([10, 10, 0], [11, 10, 0], [10, 11, 0])

    out = conservative_interpolate(src, tgt, default_value=-1.0)
    assert out.cell_data["f"][0][0] == pytest.approx(-1.0)


@needs_core
def test_conflict_policies():
    src = _one_triangle([0, 0, 0], [1, 0, 0], [0, 1, 0])
    src.cell_data["f"] = [np.array([2.0])]
    tgt = _one_triangle([0, 0, 0], [1, 0, 0], [0, 1, 0])
    tgt.cell_data["f"] = [np.array([99.0])]

    with pytest.raises(ValueError):
        conservative_interpolate(src, tgt)

    out_ow = conservative_interpolate(src, tgt, on_conflict="overwrite")
    assert out_ow.cell_data["f"][0][0] == pytest.approx(2.0)

    out_sfx = conservative_interpolate(src, tgt, on_conflict="suffix")
    assert "f_interp" in out_sfx.cell_data
    assert out_sfx.cell_data["f"][0][0] == pytest.approx(99.0)


@needs_core
def test_unknown_array_raises():
    src = _one_triangle([0, 0, 0], [1, 0, 0], [0, 1, 0])
    tgt = _one_triangle([0, 0, 0], [1, 0, 0], [0, 1, 0])
    with pytest.raises(ValueError):
        conservative_interpolate(src, tgt, arrays=["nope"])


@needs_core
def test_mismatched_topological_dimension_raises():
    src = _one_triangle([0, 0, 0], [1, 0, 0], [0, 1, 0])
    src.cell_data["f"] = [np.array([1.0])]
    tgt = meshioplusplus.Mesh(
        np.array([[0, 0, 0], [1, 0, 0], [0, 1, 0], [0, 0, 1]], float),
        [("tetra", np.array([[0, 1, 2, 3]]))],
    )
    with pytest.raises(ValueError):
        conservative_interpolate(src, tgt)


@needs_core
def test_target_geometry_and_own_data_untouched():
    src = _one_triangle([0, 0, 0], [1, 0, 0], [0, 1, 0])
    src.cell_data["f"] = [np.array([2.0])]
    tgt = _one_triangle([0, 0, 0], [1, 0, 0], [0, 1, 0])
    tgt.cell_data["g"] = [np.array([42.0])]

    out = conservative_interpolate(src, tgt)
    assert len(out.points) == len(tgt.points)
    assert out.cell_data["g"][0][0] == pytest.approx(42.0)


@needs_core
def test_ragged_polyhedron_source_and_target_are_accepted():
    src = _one_polyhedron_tet([0, 0, 0], [1, 0, 0], [0, 1, 0], [0, 0, 1])
    src.cell_data["f"] = [np.array([3.5])]
    tgt = _one_polyhedron_tet([0, 0, 0], [1, 0, 0], [0, 1, 0], [0, 0, 1])

    out = conservative_interpolate(src, tgt)
    assert out.cell_data["f"][0][0] == pytest.approx(3.5, abs=1e-9)


def test_conservative_interpolate_is_exported():
    assert "conservative_interpolate" in meshioplusplus.__all__
    assert meshioplusplus.conservative_interpolate is conservative_interpolate


def _tet_volumes(mesh):
    p = np.asarray(mesh.points, dtype=float)
    c = p[np.concatenate([np.asarray(b.data) for b in mesh.cells])]
    cross = np.cross(c[:, 1] - c[:, 0], c[:, 2] - c[:, 0])
    return np.einsum("ij,ij->i", cross, c[:, 3] - c[:, 0]) / 6.0


@pytest.mark.parametrize(
    "height, depth, layers",
    [(1.05928007, 1.0, 2), (1.0, 1.0, 2), (0.3, 1.0, 2), (1.0, 1.5, 3)],
)
def test_aligned_lattices_conserve_mass_exactly(height, depth, layers):
    # Found by tests/python/test_properties.py: two simplexified lattices of
    # one box whose tetrahedra share cutting planes. A source corner lying ON
    # a target face plane was left out of the clip's capping polygon, so one
    # overlap came out ~6% small and the transferred mass 0.7% short. Then a
    # corner on a target plane only up to rounding (a layer at z = 1/3) was
    # clipped as outside but capped as on-plane, losing a sliver.
    src = meshioplusplus.convert_cells(
        meshioplusplus.grid((1, 1, 1), spacing=(1.0, depth, height)), "simplexify"
    )
    if layers == 2:
        tgt_grid = meshioplusplus.grid((1, 2, 1), spacing=(1.0, depth / 2, height))
    else:
        tgt_grid = meshioplusplus.grid(
            (1, 1, layers), spacing=(1.0, depth, height / layers)
        )
    tgt = meshioplusplus.convert_cells(tgt_grid, "simplexify")
    src = meshioplusplus.Mesh(src.points, [(b.type, b.data) for b in src.cells])
    tgt = meshioplusplus.Mesh(tgt.points, [(b.type, b.data) for b in tgt.cells])
    values = np.array(
        [6.4059207, 2.77088847, 0.50563789, 0.26362359, 8.15137537, 9.13628022]
    )
    src.cell_data["m"] = [values]
    out = conservative_interpolate(src, tgt, arrays=["m"])
    mass_in = float(values @ _tet_volumes(src))
    mass_out = float(np.concatenate(out.cell_data["m"]) @ _tet_volumes(out))
    assert mass_out == pytest.approx(mass_in, rel=1e-12)
